Introduce `StakePoolState` to be used instead of `PoolParams`

aniketd opened the pull request:
# Description

Resolves #4989

# Checklist

- [x] Commits in meaningful sequence and with useful messages.
- [x] Tests added or updated when needed.
- [x] `CHANGELOG.md` files updated for packages with externally visible changes.  
      **NOTE: _New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd)).
- [ ] ~~Versions updated in `.cabal` and `CHANGELOG.md` files when necessary, according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).~~
- [ ] ~~Version bounds in `.cabal` files updated when necessary.  
      **NOTE: _If bounds change in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process)).~~
- [x] Code formatted (use `scripts/fourmolize.sh`).
- [x] Cabal files formatted (use `scripts/cabal-format.sh`).
- [x] CDDL files are up to date (use `scripts/gen-cddl.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) updated (use `scripts/gen-hie.sh`).
- [x] Self-reviewed the diff.

aniketd committed:
Replace PoolParams with StakePoolState in PState

lehins reviewed (changes requested):
The module move was done incorrectly resulting in the loss of git history. It needs to be done in two separate commits:
1. Move the module
2. recreate old module with re-exports and deprecations

Other than that it looks great!

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/PoolParams.hs:
```diff
@@ -1,327 +1,23 @@
+-- to provide better separation between pool registration parameters and pool state
+-- management. All types are now re-exported through "Cardano.Ledger.State" for
+-- convenient access.
+module Cardano.Ledger.PoolParams
```
Naturally, this move will need to be recorded in the changelog, but I am sure you planned to do that already :wink:

aniketd committed:
Move PoolParams to State.StakePool

aniketd committed:
Fix build: moving PoolParams to State.StakePool

aniketd committed:
Add StakePoolState to State.StakePool

aniketd committed:
Add PoolParams module back as deprecated.

Also deprecate the API:
- Deprecate getPoolParameters in favour of getPoolState
- Reintroduce and deprecate lens epochStatePoolParamsL in favour of
epochStatePoolStateL

aniketd committed:
Add changelog

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Internal.hs:
```diff
@@ -568,7 +567,7 @@ data RatifyEnv era = RatifyEnv
   , reCommitteeState :: CommitteeState era
   , reAccounts :: Accounts era
-  , rePoolParams :: Map (KeyHash 'StakePool) PoolParams
+  , rePoolState :: Map (KeyHash 'StakePool) StakePoolState
```
Let's rename this as well, while we are at it. This will also need to be mentioned in the changelog
```suggestion
  , reStakePools :: Map (KeyHash 'StakePool) StakePoolState
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/State/CertState.hs:
```diff
@@ -225,10 +225,10 @@ lookupRewardDState DState {dsAccounts} cred = do
-  -- ^ The stake pool parameters.
-  , psFutureStakePoolParams :: !(Map (KeyHash 'StakePool) PoolParams)
-  -- ^ The future stake pool parameters.
+  { psStakePoolState :: !(Map (KeyHash 'StakePool) StakePoolState)
```
Let's rename this, since it is not just a state for a single stake pool. We can omit "state` suffix, since `PState` already stands for Pools State
```suggestion
  { psStakePools :: !(Map (KeyHash 'StakePool) StakePoolState)
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/State/CertState.hs:
```diff
@@ -225,10 +225,10 @@ lookupRewardDState DState {dsAccounts} cred = do
-  -- ^ The future stake pool parameters.
+  { psStakePoolState :: !(Map (KeyHash 'StakePool) StakePoolState)
+  -- ^ The stake pool state.
+  , psFutureStakePoolState :: !(Map (KeyHash 'StakePool) StakePoolState)
```
Ditto
```suggestion
  , psFutureStakePools :: !(Map (KeyHash 'StakePool) StakePoolState)
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/State/CertState.hs:
```diff
@@ -225,10 +225,10 @@ lookupRewardDState DState {dsAccounts} cred = do
+  { psStakePoolState :: !(Map (KeyHash 'StakePool) StakePoolState)
+  -- ^ The stake pool state.
+  , psFutureStakePoolState :: !(Map (KeyHash 'StakePool) StakePoolState)
+  -- ^ The future stake pool state.
```
```suggestion
  -- ^ The state of future stake pools.
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/State/CertState.hs:
```diff
@@ -225,10 +225,10 @@ lookupRewardDState DState {dsAccounts} cred = do
-  , psFutureStakePoolParams :: !(Map (KeyHash 'StakePool) PoolParams)
-  -- ^ The future stake pool parameters.
+  { psStakePoolState :: !(Map (KeyHash 'StakePool) StakePoolState)
+  -- ^ The stake pool state.
```
```suggestion
  -- ^ The state of current stake pools.
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -1778,10 +1777,10 @@ delegateSPORewardAddressToDRep_ ::
   void $
     delegateToDRep
-      (raCredential . ppRewardAccount $ pp)
+      (raCredential . spsRewardAccount $ sps)
```
Slight simplification
```suggestion
      (raCredential $ spsRewardAccount sps)
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/SPORatifySpec.hs:
```diff
@@ -187,7 +186,7 @@ data TestData era = TestData
   , stakeNotVoted :: Coin
   , delegatees :: Map (Credential 'Staking) DRep
-  , poolParams :: Map (KeyHash 'StakePool) PoolParams
+  , poolState :: Map (KeyHash 'StakePool) StakePoolState
```
```suggestion
  , stakePools :: Map (KeyHash 'StakePool) StakePoolState
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs:
```diff
@@ -302,11 +302,11 @@ epochTransition = do
 
   -- Activate future StakePools
-  let newStakePoolParams = eval (psStakePoolParams pState0 ⨃ psFutureStakePoolParams pState0)
+  let newStakePoolState = eval (psStakePoolState pState0 ⨃ psFutureStakePoolState pState0)
```
```suggestion
  let newStakePools = eval (psStakePoolState pState0 ⨃ psFutureStakePoolState pState0)
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/API/Wallet.hs:
```diff
@@ -168,7 +168,22 @@ getPoolParameters ::
+-- | Get the /current/ registered stake pool state for a given set of
+-- stake pools. The result map will contain entries for all the given stake
+-- pools that are currently registered.
+getPoolState ::
```
For consistency sake
```suggestion
getStakePools ::
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState/Types.hs:
```diff
@@ -695,9 +694,15 @@ epochStateTreasuryL :: Lens' (EpochState era) Coin
 
+epochStatePoolStateL ::
+  EraCertState era => Lens' (EpochState era) (Map (KeyHash 'StakePool) StakePoolState)
+epochStatePoolStateL = esLStateL . lsCertStateL . certPStateL . psStakePoolStateL
```
```suggestion
epochStateStakePoolsL ::
  EraCertState era => Lens' (EpochState era) (Map (KeyHash 'StakePool) StakePoolState)
epochStateStakePoolsL = esLStateL . lsCertStateL . certPStateL . psStakePoolStateL
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Epoch.hs:
```diff
@@ -212,12 +212,12 @@ epochTransition = do
-  let PState pParams fPParams _ _ = pstate
-      ppp = eval (pParams ⨃ fPParams)
+  let PState pState fPState _ _ = pstate
+      ppp = eval (pState ⨃ fPState)
```
```suggestion
  let PState stakePools futureStakePools _ _ = pstate
      ppp = eval (pState ⨃ fPState)
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/State/Stake.hs:
```diff
@@ -84,11 +85,13 @@ class
-    , ssPoolParams = VMap.fromMap psStakePoolParams
+    , ssPoolParams =
+        VMap.fromMap $
+          Map.mapWithKey stakePoolStateToPoolParams psStakePoolState
```
Going through a list here will be more efficient, since list fusion should completely remove intermediate representation, which is not the case for a Map:
```suggestion
        VMap.fromDistinctAscListN
          (Map.size psStakePoolState)
          [(poolId, stakePoolStateToPoolParams poolId ps) | (poolId, ps) <- Map.toAscList psStakePoolState]
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/State/StakePool.hs:
```diff
@@ -0,0 +1,467 @@
+parseJsonBase16 :: Value -> Parser ByteString
+parseJsonBase16 v = do
+  s <- parseJSON v
+  case B16.decode (Char8.pack s) of
```
There is actually a bug in here that would silently convert unicode to ascii, potentially causing invalid symbols becoming valid. Let's fix it:

```suggestion
  txt <- parseJSON v
  unless (Text.isAscii txt) $ fail $ "Supplied text contains non-ASCII characters: " <> show txt
  case B16.decode (Text.encodeUtf8 txt) of
```
Could you please fix it in a separate commit.

lehins reviewed (approved):
Could use some better naming for things, but otherwise it looks great.

aniketd committed:
Address review comments

aniketd committed:
Disallow non-ascii in parseJsonBase16

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs:
```diff
@@ -283,7 +282,7 @@ data DRepPulser era (m :: Type -> Type) ans where
     -- ^ Snapshot of the proposal-deposits per reward-account-staking-credential
     , dpGlobals :: !Globals
-    , dpPoolParams :: !(Map (KeyHash 'StakePool) PoolParams)
+    , dpPoolState :: !(Map (KeyHash 'StakePool) StakePoolState)
```
```suggestion
    , dpStakePools :: !(Map (KeyHash 'StakePool) StakePoolState)
```

lehins reviewed (approved):
Beautiful!
